aGHQ.qmd

```diff
@@ -0,0 +1,1402 @@
+
+```{julia}
+#| code-fold: show
+#| output: false
+contra = let tbl = MixedModels.dataset(:contra)
+  Table(tbl; ch=tbl.livch .≠ "0")
+end
+contrasts = Dict{Symbol,Any}(
+  :urban => HelmertCoding(),
+  :ch => HelmertCoding(),
+  :dist => Grouping(),
+)
+com05 =
+  let form = @formula use ~
+      1 + urban + ch * age + age & age + (1 | dist & urban)
+    fit(MixedModel, form, contra, Bernoulli(); contrasts, nAGQ=9)
+  end
+```
+
+Extract the fixed-effects model matrix, $\bbX$, and initialize the coefficient vector, $\bbbeta$, to a copy (in case we modify it) of the estimated fixed-effects.
+
+```{julia}
+βm05 = copy(com05.β)
+showcompact(βm05)   # showcompact is defined in the first code block
+```
+
+As stated above, the `meanunitdev` function can be applied to the vectors, ${\mathbf{y}}$ and ${\bbeta}$, via dot-vectorization to produce a `Vector{NamedTuple}`, which is the typical form of a row-table.
+
+```{julia}
+rowtbl = meanunitdev.(com05.y, com05.X * βm05)
+typeof(rowtbl)
+```
+
+For display we convert the row-table to a column-table and prepend another column-table consisting of $\bby$ and $\bbeta$.
+
+```{julia}
+Table((; y=com05.y, η=com05.X * βm05), rowtbl)  # display as a Table
+```
+
+The deviance for this value of ${\bbbeta}$ in this model is the sum of the unit deviances, which we write as `sum` applied to a [generator expression](https://docs.julialang.org/en/v1/manual/arrays/#Generator-Expressions).  (In general we extract columns of a row-table with generator expressions that produce [iterators](https://docs.julialang.org/en/v1/base/collections/#lib-collections-iteration).)
+
+```{julia}
+sum(r.dev for r in rowtbl)
+```
+
+### Encapsulating the model in a struct
+
+When minimizing the deviance it is convenient to have the different components of the model encapsulated in a user-created `struct` type so we can update the parameter values and evaluate the deviance without needing to keep track of all the pieces of the model.
+
+```{julia}
+struct BernoulliGLM{T<:AbstractFloat}
+  X::Matrix{T}
+  β::Vector{T}
+  ytbl::NamedTuple{(:y, :η),NTuple{2,Vector{T}}}
+  rtbl::Vector{NamedTuple{(:μ, :dev),NTuple{2,T}}}
+end
+```
+
+We also create an *external constructor*, which is a function defined outside the struct and of the same name as the struct, that constructs and returns an object of that type.
+In this case the external constructor creates a `BernoulliGLM` from the model matrix and the response vector, after some consistency checks on the arguments passed to it.
+
+```{julia}
+#| output: false
+function BernoulliGLM(
+  X::Matrix{T},
+  y::Vector{T},
+) where {T<:AbstractFloat}
+
+  # check consistency of arguments
+  n = size(X, 1)                  # number of rows in X
+  if length(y) ≠ n || !all(v -> (iszero(v) || isone(v)), y)
+    throw(ArgumentError("y is not an $n-vector of 0's and 1's"))
+  end
+
+  # initial β from linear regression of y in {-1,1} coding
```

Comment:
why change the coding for the initial estimate?

Reply:
To make it symmetric about zero.  The way that initial values are created in `GLM.jl` and in `R`'s `glm` function essentially does the same thing by initializing the mean to values of 0.25 and 0.75 then taking the logit to get eta.  In that case the values of eta all end up being -log(3) or +log(3) but it is the same idea.   IRLS is stable and fast so the initial values for the coefficients are not terribly important.

Reply:
Can you mention that somewhere in the code/text?